### PR TITLE
Emit python code run in doctests as script

### PR DESCRIPTION
**Pull request by jigold**
This produces a python script `_build/doctest/test_hail_docs.py` with all of the code tested by the Sphinx doctest module.

### CODE REVIEW

**Comment by jigold** [comment]
Run automatically when `createDocs` is run with gradle. Does not get run if task is `createDocsNoTest`.

**Comment by tpoterba** [comment]
just to confirm, the stuff at the bottom appends the startup + cleanup, right? This looks great, thanks for tackling this!

**Comment by jigold** [comment]
Yes.

File: python/hail/docs/_ext/hail_doctest.py [hunk 1]
```diff
@@ -193,6 +194,29 @@ def summarize(self, out, verbose=None):
...
         return res
 
+    def get_code(self):
+        return self.code
+
+    def run(self, test, compileflags=None, out=None, clear_globs=True):
+        original_optionflags = self.optionflags
+        for examplenum, example in enumerate(test.examples):
+            # Merge in the example's options.
+            self.optionflags = original_optionflags
```
Comment (cseed) [inline python/hail/docs/_ext/hail_doctest.py]: You need to either reset this after each iteration (at the bottom of the loop), or after the loop, too.  It looks like now you can leave with the updates modified by the last example.

Comment (jigold) [inline python/hail/docs/_ext/hail_doctest.py]: Good catch! back to you.

File: python/hail/docs/_ext/hail_doctest.py [hunk 2]
```diff
@@ -193,6 +194,32 @@ def summarize(self, out, verbose=None):
...
```
Comment (cseed) [inline python/hail/docs/_ext/hail_doctest.py]: OK, maybe I'm being stupid, but why can't you just use an local variable for optionsflags in the loop, one for each example?  That way, you don't have to mutate self.optionflags at all.

**Comment by jigold** [comment]
@cseed You're right. Not sure why the original doctest library wrote it like that.

**Pull request merged**